src/ui/draggable/draggable-synchronizer.ts

```diff
@@ -0,0 +1,103 @@
+import { Draggable } from './draggable';
+import { setf } from '../../setf/setf';
+import { Vector } from '../../mathf/vector';
+import { MatrixService } from '../carousel/draggable-slide/matrix-service';
+
+export class DraggableSynchronizer {
+  /**
+   * Returns a singleton. `use` is passed to ensure the singleton is only
+   * actually disposed when it is done being used everywhere. This same `use`
+   * should be passed to the dispose call.
+   * @param use
+   */
+  static getSingleton(use: any): DraggableSynchronizer {
+    DraggableSynchronizer.singletonUses.add(use);
```

Comment:
Can you tell me what the case for singletonUses is?  I'm wondering if we need it.

Reply:
It's a performance/disposal measure.

A singleton for synchronizing ensures that if we have two sync'd sets of Draggables: [a, b, c] and [d, e, f], if e and b are later synchronized, then the entire set [a, b, c, d, e, f] are synchronized.

Whereas without singletons we'd have [a, b, c], [d, e, f] and [e, b] which could result in jitter.

Where the singletonUses comes in is to ensure that the singleton is not discarded as soon as one use of the Singleton is disposed.

Without tracking uses, if we have Draggables T and U, they would both call upon getSingleton() to get a DraggableSynchronizer instance. If T then disposed itself and its DraggableSynchronizer instance, U would then be using a disposed instance of DraggableSynchronizer.

Whereas with uses being tracked, the singleton isn't fully disposed until all instances that use it have indicated they are done.

This pattern repeats itself in CachedMouseTracker and the CarouselSynchronizer.

Reply:
Hmmm...does this mean we have multiple singletons?  By definition of a singleton, I don't think this is a good pattern and confused by we have multiple singletons.

Can we just not do:

new CarouselSynchronizer();

Reply:
No, it is still a single singleton. We just can't rely on any user of the singleton to know fully when all other users of the singleton are done with it in order to dispose of it. So we can either have the singleton exist in perpetuity once instantiated, or we have it track its users so it is able to self-dispose once it's last user has indicated it is no longer needed.

If it is not a singleton it cannot be aware of all instances in which a carousel was synchronized and we could have multiple synchronizers playing tug of war with a single carousel.

Reply:
Re-worked this using the new graph setup Steven suggested. Disposal is no longer called, uses are no longer tracked, instead `removeCarousel` must be called where CarouselSynchronizer is used and it self-disposes as below:

```

  /**
   * Remove the given carousel from any synced sets.
   * @param carousel
   */
  removeCarousel(carousel: Carousel) {
    const syncedCarousels = this.carouselGraph.get(carousel);
    syncedCarousels.forEach((syncedCarousel) => {
      this.carouselGraph.get(syncedCarousel).delete(carousel);
    });
    this.carouselGraph.delete(carousel);

    if (
        this.carouselGraph.size === 0 &&
        CarouselSynchronizer.singleton === this
    ) {
      CarouselSynchronizer.singleton = null;
    }
  }
  ```